In [ ]:
# Initialize Otter
import otter
grader = otter.Notebook("assignment06.ipynb")

# Assignment 06: Table Operations

Welcome to Assignment 06!  Throughout the course you will complete assignments like this one. You can't learn technical subjects without hands-on practice, so these assignments are an important part of the course.

Collaborating on labs is more than okay -- it's encouraged! You should rarely remain stuck for more than a few minutes on a question, so ask a post to the discussion board or ask your instructor for help. Explaining things is beneficial, too -- the best way to solidify your knowledge of a subject is to explain it. You should **not** just copy/paste someone else's code, but rather work together to gain understanding of the task you need to complete. 

To receive credit for this assignment, answer all questions correctly and submit before the deadline.

**Due Date:** Tuesday, July 12, 2022 @ 11:59 pm

**Collaboration Policy:** Data science is a collaborative activity. While you may talk with others about the labs, we ask that you **write your solutions individually**. If you do discuss the assignments with others **please include their names below** (it's a good way to learn your classmates' names).

**Collaborators:** 

List collaborators here.

## Today's Assignment

In today's assignment, you'll learn how to:

- manipulate tables using functions from the `datascience` package.

- use functions from the `NumPy` library.

Let's get started! Run the cell below.

In [1]:
from datascience import *
import numpy as np

## Tables

In this lab we will be talking all about **Tables**. We use tables to store all sorts of data form sports statistics to population information. If there's data you have ever been curious about, it is very likely that the Internet has a table somewhere with that data.

Tables are integral to the foundation of Data Science, and we will go over how to **query** a table. **Querying** a table is basically asking information about the table. Some examples of common queries (in English, not code):

- How many data points are there?

- Which data points have a specific characteristic?

- What is the attribute of a specific data point?

- And many more!

There are so many ways we can use tables to get information we need, and there are several existing libraries in Python that we can use to do this! In this course, we will be using the `datascience` library, and if you take Data Science classes beyond this one, you may learn many more!

### Table Creation

Let's take a look at a table in action. Python does not have any tables by default, so we can either make a new one ourselves or we can import a table from a file. First, let's see how we can make our own table from scratch:

In [2]:
# We start out with an empty Table
# Note that 'table' is capitalized, 
# and there is nothing in the parentheses
our_table = Table()
our_table

In [3]:
# Now, let's put some data in our table.
# We put the name of a column (a label)
# then a comma ','
# and then a NumPy array of that column's values
# Alternating labels and column values, we fill our table.
our_table = our_table.with_columns(
    "Department", np.array(["Data Science", "Economics", "Mathematics", "Chemistry"]),
    "Course Number", np.array([94, 1, 2, 121])
)
our_table

Department   | Course Number
Data Science | 94
Economics    | 1
Mathematics  | 2
Chemistry    | 121

### Table Attributes: `num_rows` and `num_columns`

We can ask for all sorts of information about the table itself:

<!-- BEGIN QUESTION -->

**Question 1.** List the number of rows in `our_table`.

In [4]:
our_table.num_rows # SOLUTION

4

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**Question 2.** List the number of columns in `our_table`.

In [5]:
our_table.num_columns # SOLUTION

2

<!-- END QUESTION -->

### Getting Columns: `column()` and `with_columns()`

We can also ask about the data in the table using the `column()` method. As mentioned in lecture, we can pass in a `label` or an `index` to this method. We index into the columns of a table much like we do the items of a list; 0 corresponds to the first column, 1 corresponds to the second, etc....

<!-- BEGIN QUESTION -->

**Question 3.** List the items of the `Department` column from `our_table`, using the name of the column.

In [6]:
our_table.column("Department") # SOLUTION

array(['Data Science', 'Economics', 'Mathematics', 'Chemistry'],
      dtype='<U12')

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**Question 4.** List the items of the `Department` column from `our_table`, using the position number of the column.

In [7]:
our_table.column(0) # SOLUTION

array(['Data Science', 'Economics', 'Mathematics', 'Chemistry'],
      dtype='<U12')

<!-- END QUESTION -->

Say we have a table and we want to add additional data. We can use the `with_columns()` method to do this (just like we did above)! The `with_columns` method takes inputs the exact same way as the `column()` method. We need to make sure that the columns we add to the table have the same number of rows (the length of the array we pass in) as the table, otherwise we get an error.

In [8]:
# Our table has 4 rows, so our new column needs an array with 4 items, 1 for each row
our_table_new_column = our_table.with_columns("Number of Students", np.array([21, 905, 209, 63]))
our_table_new_column

Department   | Course Number | Number of Students
Data Science | 94            | 21
Economics    | 1             | 905
Mathematics  | 2             | 209
Chemistry    | 121           | 63

In [9]:
our_table_bad_column = our_table_new_column.with_columns("Too Few Rows", np.array([1, 2, 3]))
our_table_bad_column

ValueError: Column length mismatch. New column does not have the same number of rows as table.

Note the error message here. This may be a common mistake at first, so if you see this error message, check the number of items in your column arrays.

In [ ]:
# This is our final table.
# You may use the cell below the output
# to explore the table and see what you 
# can do with it so far.
our_table_new_column

### Loading a Table

Although creating our own tables by hand can be useful, more often than not the data we want to work with is far too large to be able to type out by hand. More commonly, we load datasets in from other sources using the `Table.read_table()` method. We can pass in a **file path** to this method and it will load that data into a table we can use in Python!

Let's see how this works using the file `"data/ncsu_football.csv"`:

In [4]:
file_table = Table.read_table("data/ncsu_football.csv")
file_table.show(5)

### Viewing a Table: `show()`

The use of the `show()` method **displays** the first n rows of a table. Like `print()` this does not return a value, it just displays the value to us at the end of a cell.

### Excluding Columns: `drop()`

We now have information about NC State's Football's seasons since statistics were kept. Because this file was pulled from the internet, it may have some data in it that we are not interested in, like the rows with a bunch of `nan` values (`nan` means "Not a number", and it is commonly used to indicate there is no value there).

**Do not always necessarily remove all columns with several `NaN` values from a table.** There may be a reason why the values are not present, but for a this exercise we don't need to worry about it.*

We can use the `drop()` method to remove columns like this from the table. Let's drop the `Notes` column:

**Question 5.** Drop the `Notes` column and save the new `Table` to `file_table_no_notes`.

In [5]:
file_table_no_notes = file_table.drop("Notes") # SOLUTION
file_table_no_notes.show(5)

In [ ]:
grader.check("q5")

**Question 6.** Let's also drop the `AP Pre`, `AP High`, `AP Post`, `SRS` and `SOS` columns from the table. These are statistics specific to college football, and they are not important for what we're doing. `drop()` can take in as many columns as you need, and it will drop them all from the table. Save this `Table` to `file_table_improved_columns`.

In [7]:
file_table_improved_columns = file_table_no_notes.drop("AP Pre", "AP High", "AP Post", "SRS", "SOS") # SOLUTION
file_table_improved_columns.show(5)

In [ ]:
grader.check("q6")

## Number of Years

We can see how many years this table covers by asking about how many rows the table has.

**Questiion 7.** Assign the variable `file_table_rows` to the number of rows in `file_table_improved_columns`. You should not write an integer, rather use one of the table attributes we have talked about so far to **calculate** the number of rows.

In [13]:
file_table_rows = file_table_improved_columns.num_rows # SOLUTION
file_table_rows

114

In [ ]:
grader.check("q7")

Using this value, we can calculate the first year in `file_table_improved_columns` without looking at it.

In [15]:
# The Table covers up until 2021, so subtracting the number of rows gives 
# us the first year NOT in the table. We add one to the result to get the
# first year in the table.
print("The first year in this table is:", 2021-file_table_rows+1)

The first year in this table is: 1908


### Querying

Let's try querying our new table. Let's see what conferences NC State has played in during its history using the `Conf` column:

In [16]:
conference_list = file_table_improved_columns.column("Conf")
conference_list

array(['ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC',
       'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'ACC', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern',
       'Southern', 'Southern', 'Southern', 'Southern', 'Southern', 'Ind',
       'Ind', 'Ind', 'Ind', 

As you can see, this list looks long and repetitive, but we can use the `np.unique` method to tell us all the conferences only once as they appear:

In [17]:
np.unique(conference_list)

array(['ACC', 'Ind', 'Southern'], dtype='<U8')

### Picking columns: `select()`

It appears that there are also several other columns that we are not very interested in. Instead of dropping several columns, we can use the `select()` method to grab only the columns we want. In this case, we only want to keep the `Year`, `W`, `L`, `T`, and `Pct`,  columns.

**Question 8.** Use the `select()` method to grab only the `Year`, `W`, `L`, `T`, and `Pct`,  columns. Save the `Table` to `football_table`.

In [18]:
football_table = file_table_improved_columns.select("Year", "W", "L", "T", "Pct") # SOLUTION
football_table

Year | W    | L    | T    | Pct
2021 | 9    | 3    | nan  | 0.75
2020 | 8    | 4    | 0    | 0.667
2019 | 4    | 8    | 0    | 0.333
2018 | 9    | 4    | 0    | 0.692
2017 | 9    | 4    | 0    | 0.692
2016 | 7    | 6    | 0    | 0.538
2015 | 7    | 6    | 0    | 0.538
2014 | 8    | 5    | 0    | 0.615
2013 | 3    | 9    | 0    | 0.25
2012 | 7    | 6    | 0    | 0.538
... (104 rows omitted)

In [ ]:
grader.check("q8")

In [20]:
# Note that our file_table table is still in tact after this.
file_table_improved_columns

Rk   | Year | Conf | W    | L    | T    | Pct   | W.1  | L.1  | T.1  | Pct.1 | Coach(es)                          | Bowl
2    | 2021 | ACC  | 9    | 3    | nan  | 0.75  | 6    | 2    | 0    | 0.75  | Dave Doeren (9-3)                  | nan
3    | 2020 | ACC  | 8    | 4    | 0    | 0.667 | 7    | 3    | 0    | 0.7   | Dave Doeren (8-4)                  | Gator Bowl-L
4    | 2019 | ACC  | 4    | 8    | 0    | 0.333 | 1    | 7    | 0    | 0.125 | Dave Doeren (4-8)                  | nan
5    | 2018 | ACC  | 9    | 4    | 0    | 0.692 | 5    | 3    | 0    | 0.625 | Dave Doeren (9-4)                  | Gator Bowl-L
6    | 2017 | ACC  | 9    | 4    | 0    | 0.692 | 6    | 2    | 0    | 0.75  | Dave Doeren (9-4)                  | Sun Bowl-W
7    | 2016 | ACC  | 7    | 6    | 0    | 0.538 | 3    | 5    | 0    | 0.375 | Dave Doeren (7-6)                  | Independence Bowl-W
8    | 2015 | ACC  | 7    | 6    | 0    | 0.538 | 3    | 5    | 0    | 0.375 | Dave Doeren (7-6)                  | Belk Bowl-L
9    | 2014 | ACC  | 8    | 5    | 0    | 0.615 | 3    | 5    | 0    | 0.375 | Dave Doeren (8-5)                  | St. Petersburg Bowl-W
10   | 2013 | ACC  | 3    | 9    | 0    | 0.25  | 0    | 8    | 0    | 0     | Dave Doeren (3-9)                  | nan
11   | 2012 | ACC  | 7    | 6    | 0    | 0.538 | 4    | 4    | 0    | 0.5   | Tom O'Brien (7-5) Dana Bible (0-1) | Music City Bowl-L
... (104 rows omitted)

### Changing Column Labels: `relabeled()`

Some of these columns have labels that may not be best for what they store. Let's change the column labels to the following:

- Year: Stay the same

- W -> Wins

- L -> Losses

- T -> Ties

- Pct -> Winning Percentage

We can rename column labels using the `relabeled()` method. You have the choice to only relabel one column or you can relabel several at once. To change the names of multiple columns, we pass in an array of the old names and an array of the new names as the 2 inputs to `relabeled()`:

**Note:** There is another method `relabel()` which changes the original table without an `=`. **Be careful** using this method as it can change your data when you may not want to.

**Question 9.** Change the names of the columns using the `.relabeled()` method. Use the table below to change the column labels.

|**Old Label**|**New Label**|
|:------------|:------------|
|W|Wins|
|L|Losses|
|T|Ties|
|Pct|Winning Percentage|

In [21]:
old_names = np.array(["W", "L", "T", "Pct"]) # SOLUTION
new_names = np.array(["Wins", "Losses", "Ties", "Winning Percentage"]) # SOLUTION

football_table_relabeled = football_table.relabeled(old_names, new_names)
football_table_relabeled.show(5)

In [ ]:
grader.check("q9")

### Asking Questions

Now that we have the table we want, let's try to write some code that tells us some information about NC State Football's wins. Let's write some queries that can help us answer these 3 questions. The first question has been given to you, but let's write the other 2!

- What is the most wins NC State has ever had in one season?

- How many games has NC State ever lost?

- What is the average amount of games NC State wins every year?

**Remember:** You do not need to calculate the answers to these questions by hand, you should be writing queries to have Python do all the calculation for you.

**Question 10.** What is the most wins NC State has ever had in one season?

**Note:** Question 10. is done for you.

In [24]:
most_wins_ever = np.max(football_table_relabeled.column("Wins"))
most_wins_ever

11

Let's break down this query and see what it does. First, we ask for the `Wins` column of `football_table_relabeled`, which gives us access to the win total from every season. We then use the `np.max` method to find the maximum value in this array, which ultimately tells us the most wins NC State Football has even had in any one season.

Let's use similar queries to answer the other 2 questions:

**Question 11.** How many games has NC State ever lost?

**Hint:** `np.sum` might be helpful.

In [25]:
games_lost_alltime = np.sum(football_table_relabeled.column("Losses")) # SOLUTION
games_lost_alltime

565

In [ ]:
grader.check("q11")

**Question 12.** What is the average amount of games NC State wins every year?

**HInt:** `np.average` might be helpful.

In [27]:
average_wins = np.average(football_table_relabeled.column("Wins")) # SOLUTION
average_wins

5.184210526315789

In [ ]:
grader.check("q12")

### Sorting a Column: `sort()`

We will now introduce a new table method: `sort()`. `sort()` allows us to see a table's column values sorted by its values, from either **biggest-to-smallest** (`descending=True`) or **smallest-to-biggest** (`descending=False`).

Let's say we want to ask the question: **What is NC State's best season ever?**. There are many ways to answer the question, but you may argue that a season with the most wins or the fewest losses could be considered the best:

In [ ]:
# We can sort in descending order:
football_table_relabeled.sort("Wins", descending=True)

In [ ]:
# Or we can sort in ascending order:
football_table_relabeled.sort("Losses", descending=False)

As you can see, queries about the most wins and the fewest losses can both answer the question **What is NC State's best season ever?** in different ways. Note that the same seasons do not necessarily show up in the top of each queried table.

Yet another way to answer this question about NC State's best seasons ever is to sort by winning percentage.

**Question 13.** Assign the variable `best_win_pct` to the result of a table query sorting the table based on winning percentage.

In [36]:
best_win_pct = football_table_relabeled.sort("Winning Percentage", descending=True) # SOLUTION
best_win_pct

Year | Wins | Losses | Ties | Winning Percentage
1927 | 9    | 1      | 0    | 0.9
1913 | 6    | 1      | 0    | 0.857
1909 | 6    | 1      | 0    | 0.857
1908 | 6    | 1      | 0    | 0.857
1910 | 4    | 0      | 2    | 0.833
1967 | 9    | 2      | 0    | 0.818
1957 | 7    | 1      | 2    | 0.8
1974 | 9    | 2      | 1    | 0.792
2002 | 11   | 3      | 0    | 0.786
1944 | 7    | 2      | 0    | 0.778
... (104 rows omitted)

In [ ]:
grader.check("q13")

As you can see, many of NC State Football's best seasons are quite far in the past, only a few modern seasons even show up in any of these queries.

### Row Selection: `where()` and the `are` Predicates

The last table method we will talk about is the `where()` method. The `where()` method keeps all rows that satisfies a particular boolean condition. It takes in a column label and an `are` statement, which can be crafted using the `are` library. These are the most important `are` library methods, but there are many more if you would like to investigate: [Explore the 'are' library here.](http://data8.org/datascience/predicates.html)

| Method | Input Type | Method Description |
| --- | --- | --- |
| `are.equal_to(n)` | number | Is the value from the column equal to `n`? |
| `are.above(n)` | number | Is the value from the column above `n`? |
| `are.above_or_equal_to(n)` | number | Is the value from the column above or equal to `n`? |
| `are.below(n)` | number | Is the value from the column below `n`? |
| `are.below_or_equal_to(n)` | number | Is the value from the column below or equal `n`? |
| `are.containing(s)` | string | Is `s` contained in the string value from the given column? |
| `are.containined_in(s)` | string | Is the string value from the given column contained in `s`? |

Adding a `not_` in front of all of these methods makes each method do the opposite of what it does (ex: `are.not_equal_to(n)`).

For example, if we only wanted to see the NC State Football seasons where they had a tie, we could use the `where()` method combined with an `are` method:

In [29]:
football_table_relabeled.where("Ties", are.equal_to(0))

Year | Wins | Losses | Ties | Winning Percentage
2020 | 8    | 4      | 0    | 0.667
2019 | 4    | 8      | 0    | 0.333
2018 | 9    | 4      | 0    | 0.692
2017 | 9    | 4      | 0    | 0.692
2016 | 7    | 6      | 0    | 0.538
2015 | 7    | 6      | 0    | 0.538
2014 | 8    | 5      | 0    | 0.615
2013 | 3    | 9      | 0    | 0.25
2012 | 7    | 6      | 0    | 0.538
2011 | 8    | 5      | 0    | 0.615
... (71 rows omitted)

Or if we wanted to see NC State's worst seasons where their winning percentage was worse than .500, we can use a similarly query:

In [30]:
football_table_relabeled.where("Winning Percentage", are.below(.5))

Year | Wins | Losses | Ties | Winning Percentage
2019 | 4    | 8      | 0    | 0.333
2013 | 3    | 9      | 0    | 0.25
2009 | 5    | 7      | 0    | 0.417
2008 | 6    | 7      | 0    | 0.462
2007 | 5    | 7      | 0    | 0.417
2006 | 3    | 9      | 0    | 0.25
2004 | 5    | 6      | 0    | 0.455
1996 | 3    | 8      | 0    | 0.273
1995 | 3    | 8      | 0    | 0.273
1987 | 4    | 7      | 0    | 0.364
... (41 rows omitted)

Again you can see that NC State Football (especially recently) has had some rough seasons.

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

When done exporting, download the .zip file by `SHIFT`-clicking on the file name and selecting **Save Link As**. Or, find the .zip file in the left side of the screen and right-click and select **Download**. You'll submit this .zip file for the assignment in Canvas to Gradescope for grading.

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False)